### DISCUSSION TITLE
best practice for custom http responses

### DISCUSSION THREAD

**Discussion by TomerSalton** [loopbackio/loopback-next#7174] [Q&A]
Hey,

I want to change the HTTP response (both errors and message) that is returned by the server.
Two examples:
1. If a user tries to send a PUT request in order to create an item, I want to return him a "ERROR 501 Not Implemented" response.
2. If a user sends a POST request that requires a lot of processing power, I want to return HTTP 202 (Created).

Would you be able to provide me the best practice of implementing those HTTP responses customizations?

Thanks alot for your help,
Tomer.

**Comment by dhmlau** [comment]
@TomerSalton, do you mean to change the HTTP response outside the controller? 
If yes, could you please take a look at the discussion thread in https://github.com/strongloop/loopback-next/issues/1867? especially [this comment](https://github.com/strongloop/loopback-next/issues/1867#issuecomment-434247807) and [this comment](https://github.com/strongloop/loopback-next/issues/1867#issuecomment-456206590). Thanks.

**Comment by achrinza** [comment]
Is there any plan to do first-class support on custom HTTP response (not just error-handling)?

[I've followed the solution on StackOverflow](https://stackoverflow.com/questions/54655257/how-to-set-a-different-http-status-in-loopback-4). However, this meant that the repository now expects an `@loopback/rest.Response`, on which how to mock is not documented.

Otherwise, is there a way to create controller unit tests?

**Comment by raymondfeng** [comment]
See an example at https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/__tests__/unit/writer.unit.ts. `stubExpressContext` is the helper.

**Comment by stale** [comment]
This issue has been marked stale because it has not seen activity within six months. If you believe this to be in error, please contact one of the code owners, listed in the `CODEOWNERS` file at the top-level of this repository. This issue will be closed within 30 days of being stale.
